src/Simulation/Simulators/SparseSimulator/README.md

```diff
@@ -0,0 +1,155 @@
+
+Since the hash map does not allow modifications while iterating through it, this is the only way to implement most gates.
+
+Finally, the `BasicQuantumState` class needs to implement the same function virtually, so the `SparseSimulator` class can call it.
+
+## SparseSimulator (SparseSimulator.h)
+The `SparseSimulator` class will also need a wrapper function for the new gate. This will call `_quantum_state->your_new_gate(args...);`. However, `SparseSimulator` also manages gate queues, so you will need to decide how the new gate commutes with existing ones. The simplest approach will be to start by calling `_execute_queued_ops();`, which will flush all the gate queues and allow you to run your gate without needing any commutation relations.
+
+## C API (capi.cpp)
+Within capi.cpp, you need to add a new wrapper to call the function in `SparseSimulator`. The first argument must be an unsigned integer, representing the ID of the simulator, so you will call it as 
+`getSimulator(sim_id)->your_new_gate(\# args \#);`
+For more complicated arguments from C\#, the other functions provide a template. For consistency, append "\_cpp" to the end of the function name in capi.cpp.
+
+
+## SparseSimulator (SparseSimulator.cs)
+Here the C\# code will call your new gate. The wrapper must go in the `SparseSimulatorProcessor` class, and must have the following form:
+`private static extern void your_new_gate_cpp(uint sim, \# args\#);`
+`public override void Your_new_gate(\#args\#)
+{
+    your_new_gate_cpp(Id, \# arguments, parsed into a format that C++ can read\#);
+}`
+Here `Id` is an internal variable for the simulator's ID, which must be passed to C++.
+
+## (Optional) Q\# Code
+If you are implementing a gate that Q\# already expects (i.e., it is initialized in the `QuantumProcessorBase` class) then the previous steps will be enough. However, if you want to create an entirely new gate, you will need to create a Q\# operation to call it.
+
+"Probes.qs" and "Probes.cs" provide a template for how this code will look. In Q\#, declare the operation `YourNewGate` with code `body intrinsic;`. Then in a separate C\# file, use the following template:
+`public partial class YourNewGate
+{
+    public class Native : YourNewGate
+    {
+        private SparseSimulator sim = null;
+        public Native(IOperationFactory m) : base(m)
+        {
+            sim = m as SparseSimulator;
+        }
+        public override Func<\#input types from Q\#\#, \#output types to Q\#\#> __Body__ => sim == null ? base.__Body__ : (args) => {
+            return sim.Your_new_gate(args.Item1, args.Item2,...);
+        };
+    }
+}`
```

Comment:
(Use triple tick to start and end the code block, and nothing else in the triple-tick lines.
On the right of the opening triple tick you can specify the language name to use the syntax highlighting.
See details in the [Fenced Code Blocks](https://www.markdownguide.org/extended-syntax/#fenced-code-blocks) and [Syntax Highlighting](https://www.markdownguide.org/extended-syntax/#syntax-highlighting))
```suggestion
```c#
public partial class YourNewGate
{
    public class Native : YourNewGate
    {
        private SparseSimulator sim = null;
        public Native(IOperationFactory m) : base(m)
        {
            sim = m as SparseSimulator;
        }
        public override Func<\#input types from Q\#\#, \#output types to Q\#\#> __Body__ => sim == null ? base.__Body__ : (args) => {
            return sim.Your_new_gate(args.Item1, args.Item2,...);
        };
    }
}
\```
```

Reply:
Addressed it in https://github.com/microsoft/qsharp-runtime/pull/769